File: test/SaltarelleProjectTests/CoreSubset/Core/Utility/PointUtil.cs
// ---------------------------------------------------------------------------------------------------------------------
// <copyright file="PointUtil.cs" company="Tableau Software">
//   This file is the copyrighted property of Tableau Software and is protected by registered patents and other
//   applicable U.S. and international laws and regulations.
//
//   Unlicensed use of the contents of this file is prohibited. Please refer to the NOTICES.txt file for further details.
// </copyright>
// ---------------------------------------------------------------------------------------------------------------------

namespace Tableau.JavaScript.Vql.Core
{
    using System;
    using jQueryApi;
    using Tableau.JavaScript.Vql.TypeDefs;

    /// <summary> This class provides methods that operate on Point records. </summary>
    public static class PointUtil
    {
        //// ===========================================================================================================
        //// Methods
        //// ===========================================================================================================

        public static Point FromPresModel(PointPresModel pointPM)
        {
            if (Script.IsNullOrUndefined(pointPM)) { return null; }
            return new Point(pointPM.X, pointPM.Y);
        }

        public static PointPresModel ToPresModel(Point pt)
        {
            if (Script.IsNullOrUndefined(pt)) { return null; }
            PointPresModel pointPM = new PointPresModel();
            pointPM.X = pt.X;
            pointPM.Y = pt.Y;
            return pointPM;
        }

        /// <summary>
        /// Converts a <see cref="jQueryPosition"/> to a <see cref="Point"/>.
        /// </summary>
        /// <param name="position">The position to convert.</param>
        /// <returns>A new <see cref="Point"/> representing the same coordinates as <paramref name="position"/>.</returns>
        public static Point FromPosition(jQueryPosition position)
        {
            return new Point(position.Left.RoundToInt(), position.Top.RoundToInt());
        }

        /// <summary>
        /// Adds first and second and returns the result. Normally this would be an operator overload, but Script#
        /// doesn't support them.
        /// </summary>
        /// <param name="first">The first point.</param>
        /// <param name="second">The second point.</param>
        /// <returns>The sum of first and second.</returns>
        public static Point Add(Point first, Point second)
        {
            if (Script.IsNullOrUndefined(first) || Script.IsNullOrUndefined(second))
            {
                return first;
            }

            return new Point(first.X + second.X, first.Y + second.Y);
        }

        /// <summary>
        /// Subtracts first - second and returns the result. Normally this would be an operator overload, but Script#
        /// doesn't support them.
        /// </summary>
        /// <param name="first">The first point.</param>
        /// <param name="second">The second point.</param>
        /// <returns>The difference of first - second.</returns>
        public static Point Subtract(Point first, Point second)
        {
            return new Point(first.X - second.X, first.Y - second.Y);
        }

        /// <summary>
        /// Gets the distance between the first and second points.
        /// </summary>
        /// <param name="first">The first point.</param>
        /// <param name="second">The second point.</param>
        /// <returns>The distance between first and second</returns>
        public static double Distance(Point first, Point second)
        {
            Param.VerifyValue(first, "first");
            Param.VerifyValue(second, "second");

            int diffX = first.X - second.X;
            int diffY = first.Y - second.Y;

            return Math.Sqrt((diffX * diffX) + (diffY * diffY));
        }

        /// <summary>
        /// Quick distance test between two points
        /// </summary>
        /// <param name="first">The first point.</param>
        /// <param name="second">The second point.</param>
        /// <param name="distance">Maximum distance between the two points.</param>
        /// <returns>true if points are less than or equal to the specified distance</returns>
        public static bool IsWithinDistance(Point first, Point second, long distance)
        {
            int diffX = first.X - second.X;
            int diffY = first.Y - second.Y;
            return ((diffX * diffX) + (diffY * diffY)) <= (distance * distance);
        }

        /// <summary>
        /// Tests if two points are equal.
        /// </summary>
        /// <param name="p">The first point to test</param>
        /// <param name="p2">The second point to test</param>
        /// <returns>True if p and p2 are not null and p.x,y equal p2.x,y</returns>
        public static bool Equals(Point p, Point p2)
        {
            return Script.IsValue(p) && Script.IsValue(p2) && p2.X == p.X && p2.Y == p.Y;
        }
    }

    /// <summary> This class provides methods that operate on PointF records. </summary>
    public static class PointFUtil
    {
        /// <summary>
        /// Subtracts first - second and returns the result. Normally this would be an operator overload, but Script#
        /// doesn't support them.
        /// </summary>
        /// <param name="first">The first point.</param>
        /// <param name="second">The second point.</param>
        /// <returns>The difference of first - second.</returns>
        public static PointF Subtract(PointF first, PointF second)
        {
            return new PointF(first.X - second.X, first.Y - second.Y);
        }

        public static PointF TimesScalar(Point p, float scalar)
        {
            return new PointF(p.X * scalar, p.Y * scalar);
        }

        public static Point Round(PointF p)
        {
            return new Point(Math.JsRound(p.X), Math.JsRound(p.Y));
        }
    }
}

File: E2ETestResults/CoreSubset/Core/Utility/PointUtil.ts
import { DoubleUtil } from './DoubleUtil';

import { jQueryPosition } from 'jQuery';

import 'mscorlib';

import { Param } from './Param';

import { Point, PointD } from './Point';

import { PointPresModel } from 'TypeDefs';

/**
 * This class provides methods that operate on Point records.
 */
export class PointUtil {
  public static fromPresModel(pointPM: PointPresModel): Point {
    if (ss.isNullOrUndefined(pointPM)) {
      return null;
    }
    return new Point(pointPM.x, pointPM.y);
  }

  public static toPresModel(pt: Point): PointPresModel {
    if (ss.isNullOrUndefined(pt)) {
      return null;
    }
    let pointPM: PointPresModel = new PointPresModel();
    return pointPM;
  }

  /**
   * Converts a {@link jQueryPosition} to a {@link Point}.
   * @param position The position to convert.
   * @returns A new {@link Point} representing the same coordinates as .
   */
  public static fromPosition(position: jQueryPosition): Point {
    return new Point(DoubleUtil.roundToInt(position.left), DoubleUtil.roundToInt(position.top));
  }

  /**
   * Converts a {@link PointD} to a {@link Point}.
   */
  public static fromPointD(pointD: PointD): Point {
    return new Point(DoubleUtil.roundToInt(pointD.x), DoubleUtil.roundToInt(pointD.y));
  }

  /**
   * Adds first and second and returns the result. Normally this would be an operator overload, but Script#
   * doesn't support them.
   * @param first The first point.
   * @param second The second point.
   * @returns The sum of first and second.
   */
  public static add(first: Point, second: Point): Point {
    if (ss.isNullOrUndefined(first) || ss.isNullOrUndefined(second)) {
      return first;
    }
    return new Point(first.x + second.x, first.y + second.y);
  }

  /**
   * Subtracts first - second and returns the result. Normally this would be an operator overload, but Script#
   * doesn't support them.
   * @param first The first point.
   * @param second The second point.
   * @returns The difference of first - second.
   */
  public static subtract(first: Point, second: Point): Point {
    return new Point(first.x - second.x, first.y - second.y);
  }

  /**
   * Multiplys first * second and returns the product of the X components and Y components.
   * @param first The first point.
   * @param second The second point.
   * @returns The product of first * second.
   */
  public static multiply(first: Point, second: Point): Point {
    return new Point(first.x * second.x, first.y * second.y);
  }

  /**
   * Gets the distance between the first and second points.
   * @param first The first point.
   * @param second The second point.
   * @returns The distance between first and second
   */
  public static distance(first: Point, second: Point): number {
    Param.verifyValue(first, 'first');
    Param.verifyValue(second, 'second');
    let diffX: number = first.x - second.x;
    let diffY: number = first.y - second.y;
    return Math.sqrt((diffX * diffX) + (diffY * diffY));
  }

  /**
   * Quick distance test between two points
   * @param first The first point.
   * @param second The second point.
   * @param distance Maximum distance between the two points.
   * @returns true if points are less than or equal to the specified distance
   */
  public static isWithinDistance(first: Point, second: Point, distance: number): boolean {
    let diffX: number = first.x - second.x;
    let diffY: number = first.y - second.y;
    return ((diffX * diffX) + (diffY * diffY)) <= (distance * distance);
  }

  /**
   * Tests if two points are equal.
   * @param p The first point to test
   * @param p2 The second point to test
   * @returns True if p and p2 are not null and p.x,y equal p2.x,y
   */
  public static equals(p: Point, p2: Point): boolean {
    return ss.isValue(p) && ss.isValue(p2) && p2.x === p.x && p2.y === p.y;
  }

  public static timesScalar(p: Point, scalar: number): Point {
    return new Point(p.x * scalar, p.y * scalar);
  }
}

/**
 * This class provides methods that operate on PointD records.
 */
export class PointDUtil {
  /**
   * Subtracts first - second and returns the result. Normally this would be an operator overload, but Script#
   * doesn't support them.
   * @param first The first point.
   * @param second The second point.
   * @returns The difference of first - second.
   */
  public static subtract(first: PointD, second: PointD): PointD {
    return new PointD(first.x - second.x, first.y - second.y);
  }

  public static timesScalar(p: Point, scalar: number): PointD {
    return new PointD(p.x * scalar, p.y * scalar);
  }

  public static round(p: PointD): Point {
    return new Point(Math.round(p.x), Math.round(p.y));
  }
}
